<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Jonathan Chang" />


<title>Comparative methods with the Fish Tree of Life</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>

</head>

<body>




<h1 class="title toc-ignore">Comparative methods with the Fish Tree of Life</h1>
<h4 class="author"><em>Jonathan Chang</em></h4>



<p>To demonstrate the functionality of the <code>fishtree</code> package and how it integrates well with the rest of the R phylogenetics ecosystem, this vignette will walk you through a simple comparative analysis.</p>
<div id="getting-and-cleaning-data" class="section level2">
<h2>Getting and cleaning data</h2>
<p>A common hypothesis tested in comparative methods is whether habitat shifts drive rates of diversification in various groups. For example, <a href="https://doi.org/10.1111/jeb.12112">Santini et al. (2013)</a> tested, among other things, whether reef-associated pufferfishes enjoyed faster rates of speciation compared to their non-reef relatives.</p>
<p>First, load the <code>fishtree</code> package and download the subset of the Fish Tree of Life corresponding to the taxon of interest. To make things more interesting we’ll work on the entire order (<a href="https://fishtreeoflife.org/taxonomy/order/Tetraodontiformes/">Tetraodontiformes</a>) rather than the family in the 2013 study.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">library</span>(fishtree)</a>
<a class="sourceLine" id="cb1-2" title="2"></a>
<a class="sourceLine" id="cb1-3" title="3">tree &lt;-<span class="st"> </span><span class="kw">fishtree_phylogeny</span>(<span class="dt">rank =</span> <span class="st">&quot;Tetraodontiformes&quot;</span>)</a>
<a class="sourceLine" id="cb1-4" title="4"></a>
<a class="sourceLine" id="cb1-5" title="5"><span class="kw">plot</span>(tree, <span class="dt">show.tip.label =</span> <span class="ot">FALSE</span>, <span class="dt">no.margin =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p>Next, we need to get habitat data and associate it with our phylogeny. The <code>rfishbase</code> package (<a href="https://doi.org/10.1111/j.1095-8649.2012.03464.x">Boettiger et al. 2012</a>) has a variety of convenient functions to access data recorded by the Fishbase editors. Load the <code>rfishbase</code> package and retrieve the relevant data in the <code>DemersPelag</code> field, which identifies whether a species is reef-associated or not, among other things.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw">library</span>(rfishbase)</a>
<a class="sourceLine" id="cb2-2" title="2"></a>
<a class="sourceLine" id="cb2-3" title="3">tips &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;_&quot;</span>, <span class="st">&quot; &quot;</span>, tree<span class="op">$</span>tip.label, <span class="dt">fixed =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb2-4" title="4"></a>
<a class="sourceLine" id="cb2-5" title="5">fb_results &lt;-<span class="st"> </span><span class="kw">species</span>(<span class="dt">species_list =</span> tips, <span class="dt">fields =</span> <span class="kw">c</span>(<span class="st">&quot;Species&quot;</span>, <span class="st">&quot;DemersPelag&quot;</span>))</a>
<a class="sourceLine" id="cb2-6" title="6">fb_results &lt;-<span class="st"> </span>fb_results[<span class="op">!</span><span class="kw">is.na</span>(fb_results<span class="op">$</span>Species), ]</a>
<a class="sourceLine" id="cb2-7" title="7"><span class="kw">head</span>(fb_results)</a></code></pre></div>
<p>Note that we had to replace the underscores in the tip labels with spaces. This is a common source of errors, so if your analyses don’t seem to work correctly always check whether the functions you’re using expect underscores or spaces.</p>
<p>There’s a lot of data in the <code>DemersPelag</code> field, but we only want to know if the species is reef-associated or not.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">reef &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">tip =</span> <span class="kw">gsub</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;_&quot;</span>, fb_results<span class="op">$</span>Species),</a>
<a class="sourceLine" id="cb3-2" title="2">                   <span class="dt">is_reef =</span> <span class="kw">as.numeric</span>(fb_results<span class="op">$</span>DemersPelag <span class="op">==</span><span class="st"> &quot;reef-associated&quot;</span>))</a>
<a class="sourceLine" id="cb3-3" title="3"><span class="kw">head</span>(reef)</a></code></pre></div>
<p>We’ve also converted the tip labels back to underscores, since we need to ensure that the tip labels on our phylogeny match the labels on our trait data. The <code>geiger</code> package (<a href="https://doi.org/10.1093/bioinformatics/btu181">Pennell et al. 2014</a>) provides a convenient function that will perform this check. The <code>name.check</code> function expects row names on our data object, so we will do that as well.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="kw">library</span>(geiger)</a>
<a class="sourceLine" id="cb4-2" title="2"></a>
<a class="sourceLine" id="cb4-3" title="3"><span class="kw">rownames</span>(reef) &lt;-<span class="st"> </span>reef<span class="op">$</span>tip</a>
<a class="sourceLine" id="cb4-4" title="4">nc &lt;-<span class="st"> </span>geiger<span class="op">::</span><span class="kw">name.check</span>(tree, reef)</a>
<a class="sourceLine" id="cb4-5" title="5">nc</a></code></pre></div>
<p>We’ve identified a mismatch between the tree and the data. We’ll exclude the tips lacking trait data using <code>drop.tip</code>:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="kw">library</span>(ape)</a>
<a class="sourceLine" id="cb5-2" title="2">tree &lt;-<span class="st"> </span><span class="kw">drop.tip</span>(tree, nc<span class="op">$</span>tree_not_data)</a></code></pre></div>
<p>If we also had data that was not in the tree, we could exclude that using the following command, but it isn’t necessary in this case:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1">reef &lt;-<span class="st"> </span>reef[<span class="op">!</span><span class="kw">rownames</span>(reef) <span class="op">%in%</span><span class="st"> </span>nc<span class="op">$</span>data_not_tree, ]</a></code></pre></div>
<p>Confirm that we have the same number of observations in the tree and the data:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="kw">Ntip</span>(tree) <span class="op">==</span><span class="st"> </span><span class="kw">nrow</span>(reef)</a></code></pre></div>
</div>
<div id="plotting-diversification-rates" class="section level2">
<h2>Plotting diversification rates</h2>
<p>There are several other data sources available in the <code>fishtree</code> package, including speciation rates computed via the DR method (<a href="https://doi.org/10.1038/nature11631">Jetz et al. 2012</a>). Retrieve speciation rate data:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1">rates &lt;-<span class="st"> </span><span class="kw">fishtree_tip_rates</span>(<span class="dt">rank =</span> <span class="st">&quot;Tetraodontiformes&quot;</span>)</a>
<a class="sourceLine" id="cb8-2" title="2"><span class="kw">head</span>(rates)</a></code></pre></div>
<p>We’re interested in just the <code>dr</code> column, so extract that and convert spaces to underscores again. Then merge the habitat data with the speciation rate data.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">rates &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">tip =</span> <span class="kw">gsub</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;_&quot;</span>, rates<span class="op">$</span>species), <span class="dt">dr =</span> rates<span class="op">$</span>dr)</a>
<a class="sourceLine" id="cb9-2" title="2"><span class="kw">rownames</span>(rates) &lt;-<span class="st"> </span>rates<span class="op">$</span>tip</a>
<a class="sourceLine" id="cb9-3" title="3">merged &lt;-<span class="st"> </span><span class="kw">merge</span>(reef, rates)</a></code></pre></div>
<p>As a quick check our data, let’s plot histograms of the DR rate of reef and non-reef species:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">breaks &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="kw">min</span>(merged<span class="op">$</span>dr), <span class="kw">max</span>(merged<span class="op">$</span>dr), <span class="dt">length.out =</span> <span class="dv">30</span>)</a>
<a class="sourceLine" id="cb10-2" title="2"><span class="kw">hist</span>(<span class="kw">subset</span>(merged, is_reef <span class="op">==</span><span class="st"> </span><span class="dv">1</span>)<span class="op">$</span>dr, <span class="dt">col =</span> <span class="st">&quot;orange&quot;</span>, <span class="dt">density =</span> <span class="dv">20</span>, <span class="dt">angle =</span> <span class="dv">135</span>,</a>
<a class="sourceLine" id="cb10-3" title="3">     <span class="dt">breaks =</span> breaks)</a>
<a class="sourceLine" id="cb10-4" title="4"><span class="kw">hist</span>(<span class="kw">subset</span>(merged, is_reef <span class="op">==</span><span class="st"> </span><span class="dv">0</span>)<span class="op">$</span>dr, <span class="dt">col =</span> <span class="st">&quot;purple&quot;</span>, <span class="dt">density =</span> <span class="dv">20</span>, <span class="dt">angle =</span> <span class="dv">45</span>,</a>
<a class="sourceLine" id="cb10-5" title="5">     <span class="dt">breaks =</span> breaks, <span class="dt">add =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p>It seems like for the most part, the Tetraodontiformes have a low speciation rate, except for a subset of non-reef species that have a faster rate.</p>
<p>Of course, with any comparative method, it’s critical to consider the historical relationships between the species you’re examining. The following snippet of code is quite complex, but demonstrates how to draw rates onto a phylogeny using colored bars next to each tip in question.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="co"># Plot tree and extract plotting data</span></a>
<a class="sourceLine" id="cb11-2" title="2"><span class="kw">plot</span>(tree, <span class="dt">show.tip.label =</span> <span class="ot">FALSE</span>, <span class="dt">no.margin =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb11-3" title="3">obj &lt;-<span class="st"> </span><span class="kw">get</span>(<span class="st">&quot;last_plot.phylo&quot;</span>, .PlotPhyloEnv)</a>
<a class="sourceLine" id="cb11-4" title="4"></a>
<a class="sourceLine" id="cb11-5" title="5"><span class="co"># Generate a color ramp</span></a>
<a class="sourceLine" id="cb11-6" title="6">ramp &lt;-<span class="st"> </span>grDevices<span class="op">::</span><span class="kw">colorRamp</span>(<span class="kw">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="dt">bias =</span> <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb11-7" title="7">tiporder &lt;-<span class="st"> </span><span class="kw">match</span>(rates<span class="op">$</span>tip, tree<span class="op">$</span>tip.label)</a>
<a class="sourceLine" id="cb11-8" title="8">scaled_rates &lt;-<span class="st"> </span>rates<span class="op">$</span>dr <span class="op">/</span><span class="st"> </span><span class="kw">max</span>(rates<span class="op">$</span>dr, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb11-9" title="9">tipcols &lt;-<span class="st"> </span><span class="kw">apply</span>(<span class="kw">ramp</span>(scaled_rates), <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="kw">do.call</span>(rgb, <span class="kw">as.list</span>(x <span class="op">/</span><span class="st"> </span><span class="dv">255</span>)))</a>
<a class="sourceLine" id="cb11-10" title="10"></a>
<a class="sourceLine" id="cb11-11" title="11"><span class="co"># Place colored bars</span></a>
<a class="sourceLine" id="cb11-12" title="12"><span class="cf">for</span> (ii <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tiporder)) {</a>
<a class="sourceLine" id="cb11-13" title="13">    tip &lt;-<span class="st"> </span>tiporder[ii]</a>
<a class="sourceLine" id="cb11-14" title="14">    <span class="kw">lines</span>(<span class="dt">x =</span> <span class="kw">c</span>(obj<span class="op">$</span>xx[tip] <span class="op">+</span><span class="st"> </span><span class="fl">0.5</span>, obj<span class="op">$</span>xx[tip] <span class="op">*</span><span class="st"> </span><span class="fl">1.5</span> <span class="op">+</span><span class="st"> </span><span class="fl">0.5</span> <span class="op">+</span><span class="st"> </span>scaled_rates[ii]),</a>
<a class="sourceLine" id="cb11-15" title="15">          <span class="dt">y =</span> <span class="kw">rep</span>(obj<span class="op">$</span>yy[tip], <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb11-16" title="16">          <span class="dt">col =</span> tipcols[ii])</a>
<a class="sourceLine" id="cb11-17" title="17">}</a></code></pre></div>
</div>
<div id="running-an-analysis" class="section level2">
<h2>Running an analysis</h2>
<p>Let’s perform a more quantitative analysis using using <code>hisse</code>. We’ll test 4 models: a BiSSE-like model, a BiSSE-like null model, a hisse model, and the hisse 2 state null model.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1"><span class="kw">library</span>(hisse)</a></code></pre></div>
<p>The <code>hisse</code> package parameterizes things differently from <code>diversitree</code> (where BiSSE lives), so we aren’t able to exactly replicate the analyses in the Santini paper. Instead we’ll settle by ensuring that the epsilon parameter, <span class="math inline">\(\epsilon = \frac{\mu}{\lambda}\)</span> is constrained to be equal for both reef and non-reef taxa. We’ll also constrain transition rates to be equal, since it can be difficult to estimate those.</p>
<p>First we’ll construct and run the BiSSE model and the BiSSE null model:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">trans.rates.bisse &lt;-<span class="st"> </span><span class="kw">ParEqual</span>(<span class="kw">TransMatMaker</span>(<span class="dt">hidden.states =</span> <span class="ot">FALSE</span>), <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</a>
<a class="sourceLine" id="cb13-2" title="2"></a>
<a class="sourceLine" id="cb13-3" title="3">pp.bisse.full &lt;-<span class="st"> </span><span class="kw">hisse</span>(tree, reef, <span class="dt">hidden.states =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb13-4" title="4">                       <span class="dt">turnover.anc =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">0</span>), <span class="dt">eps.anc =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>),</a>
<a class="sourceLine" id="cb13-5" title="5">                       <span class="dt">trans.rate =</span> trans.rates.bisse, <span class="dt">output.type=</span><span class="st">&quot;raw&quot;</span>)</a>
<a class="sourceLine" id="cb13-6" title="6"></a>
<a class="sourceLine" id="cb13-7" title="7">pp.bisse.null &lt;-<span class="st"> </span><span class="kw">hisse</span>(tree, reef, <span class="dt">hidden.states =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb13-8" title="8">                       <span class="dt">turnover.anc =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>), <span class="dt">eps.anc =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>),</a>
<a class="sourceLine" id="cb13-9" title="9">                       <span class="dt">trans.rate =</span> trans.rates.bisse, <span class="dt">output.type=</span><span class="st">&quot;raw&quot;</span>)</a></code></pre></div>
<p>Next, we’ll run the full hisse model, save for the constrained transition rates and epsilon.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1">trans.rates.hisse &lt;-<span class="st"> </span><span class="kw">TransMatMaker</span>(<span class="dt">hidden.states =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb14-2" title="2">trans.rates.hisse &lt;-<span class="st"> </span><span class="kw">ParDrop</span>(trans.rates.hisse, <span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">8</span>,<span class="dv">10</span>))</a>
<a class="sourceLine" id="cb14-3" title="3">trans.rates.hisse &lt;-<span class="st"> </span><span class="kw">ParEqual</span>(trans.rates.hisse, <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">1</span>,<span class="dv">5</span>,<span class="dv">1</span>,<span class="dv">6</span>,<span class="dv">1</span>,<span class="dv">7</span>,<span class="dv">1</span>,<span class="dv">8</span>))</a>
<a class="sourceLine" id="cb14-4" title="4"></a>
<a class="sourceLine" id="cb14-5" title="5">pp.hisse.full &lt;-<span class="st"> </span><span class="kw">hisse</span>(tree, reef, <span class="dt">hidden.states =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb14-6" title="6">                       <span class="dt">turnover.anc=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>), <span class="dt">eps.anc=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>),</a>
<a class="sourceLine" id="cb14-7" title="7">                       <span class="dt">trans.rate=</span>trans.rates.hisse, <span class="dt">output.type=</span><span class="st">&quot;raw&quot;</span>)</a></code></pre></div>
<p>Finally, we’ll build the 2 state character independent diversification model, sometimes called CID-2. We’ll use this as our null model by forcing the visible states (reef or non-reef) to have the same net turnover rates, while permitting the hidden states to vary freely.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1">pp.hisse.null2 &lt;-<span class="st"> </span><span class="kw">hisse</span>(tree, reef, <span class="dt">hidden.states =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb15-2" title="2">                        <span class="dt">turnover.anc=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">eps.anc=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>),</a>
<a class="sourceLine" id="cb15-3" title="3">                        <span class="dt">trans.rate=</span>trans.rates.hisse, <span class="dt">output.type=</span><span class="st">&quot;raw&quot;</span>)</a></code></pre></div>
<p>We can combine all of our results into a single table for easy comparison.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1">results &lt;-<span class="st"> </span><span class="kw">list</span>(pp.bisse.full, pp.bisse.null, pp.hisse.null2, pp.hisse.full)</a>
<a class="sourceLine" id="cb16-2" title="2">aicc &lt;-<span class="st"> </span><span class="kw">sapply</span>(results, <span class="st">`</span><span class="dt">[[</span><span class="st">`</span>, <span class="st">&quot;AICc&quot;</span>)</a>
<a class="sourceLine" id="cb16-3" title="3">lnl &lt;-<span class="st"> </span><span class="kw">sapply</span>(results, <span class="st">`</span><span class="dt">[[</span><span class="st">`</span>, <span class="st">&quot;loglik&quot;</span>)</a>
<a class="sourceLine" id="cb16-4" title="4"></a>
<a class="sourceLine" id="cb16-5" title="5"><span class="kw">data.frame</span>(<span class="dt">model =</span> <span class="kw">c</span>(<span class="st">&quot;bisse_full&quot;</span>, <span class="st">&quot;bisse_null&quot;</span>, <span class="st">&quot;hisse_cid2&quot;</span>, <span class="st">&quot;hisse_full&quot;</span>), aicc, lnl)</a></code></pre></div>
<p>Summarizing the results on the basis of AICc suggests that the best supported model is a null model, where habitat has no effect on speciation rate.</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
